Практическое задание
1) Найти данные для задачи классификации или для задачи регрессии.
2) Реализовать баггинг.
3) Реализовать бустинг на тех же данных, что использовались для баггинга.
4) Сравнить результаты работы алгоритмов (время работы и качество
моделей). Сделать выводы.
5) Оформить отчет о проделанной работе.

In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import BaggingClassifier, AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, classification_report
import time

In [ ]:
data = pd.read_csv('diabetes.csv')

In [ ]:
cols_with_zero = ['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']
data[cols_with_zero] = data[cols_with_zero].replace(0, np.nan)

df_cleaned = data.dropna()
print(f"Было строк: {len(data)}, стало: {len(df_cleaned)}")

Было строк: 768, стало: 392


In [ ]:
# Разделение на признаки и целевую переменную
X = df_cleaned.drop('Outcome', axis=1)
y = df_cleaned['Outcome']

# Разделение на тренировочную и тестовую выборки
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42, stratify=y)

# **Реализация бэггинга**

In [ ]:
# Создание базовой модели (решающее дерево)
base_tree = DecisionTreeClassifier(max_depth=4, random_state=42)

# Инициализация и обучение бэггинга
start_time = time.time()
bagging = BaggingClassifier(
    estimator=base_tree,
    n_estimators=50,
    max_samples=0.8,
    max_features=0.8,
    random_state=42
)
bagging.fit(X_train, y_train)
bagging_time = time.time() - start_time

# Предсказание и оценка
bagging_pred = bagging.predict(X_test)
bagging_accuracy = accuracy_score(y_test, bagging_pred)

In [ ]:
bagging_time

0.1262226104736328

# **Реализация бустинга**

In [ ]:
# Инициализация и обучение бустинга (AdaBoost)
start_time = time.time()
boosting = AdaBoostClassifier(
    estimator=DecisionTreeClassifier(max_depth=3),
    n_estimators=50,
    learning_rate=0.8,
    random_state=42
)
boosting.fit(X_train, y_train)
boosting_time = time.time() - start_time

# Предсказание и оценка
boosting_pred = boosting.predict(X_test)
boosting_accuracy = accuracy_score(y_test, boosting_pred)


In [ ]:
boosting_time

0.12899446487426758

# **Сравнение результатов**

In [ ]:
print("баггинг")
print(classification_report(y_test, bagging_pred))

print("бустинг")
print(classification_report(y_test, boosting_pred))

баггинг
              precision    recall  f1-score   support

           0       0.85      0.87      0.86        79
           1       0.73      0.69      0.71        39

    accuracy                           0.81       118
   macro avg       0.79      0.78      0.79       118
weighted avg       0.81      0.81      0.81       118

бустинг
              precision    recall  f1-score   support

           0       0.84      0.89      0.86        79
           1       0.74      0.67      0.70        39

    accuracy                           0.81       118
   macro avg       0.79      0.78      0.78       118
weighted avg       0.81      0.81      0.81       118



# **выводы**

Качество моделей:

Бустинг показал немного лучшие результаты по всем метрикам (accuracy 74% против 73% у бэггинга).

Скорость работы:

Оба метода работают достаточно быстро на данном наборе данных.

Баггинг оказался немного быстрее